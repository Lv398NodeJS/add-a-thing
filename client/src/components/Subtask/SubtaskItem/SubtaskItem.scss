$subtask-bg-onhover: linear-gradient(to right, #ececec, #f6f6f6);
$delete-button: #f6f6f6;

.subtask-row:hover {
  background: $subtask-bg-onhover;
}

.subtask-row:hover .subtask-dropdown-toggle {
  transform: scale(1);
  transition: transform 150ms ease;
}

.subtask-dropdown-toggle {
  &,
  &:hover,
  &:link,
  &:focus,
  &:active {
    background: inherit;
    transform: scale(0);
    border: none;
    outline: none;
  }
}
.subtask-dropdown-toggle:after {
  display: none;
}
.threedots-icon {
  width: 0.9rem;
}

// Styling for checked/unchecked checkbox and its label
.custom-checkbox {
  .custom-control-input:checked {
    ~ .custom-control-label::before {
      background-color: var(--green);
      border-color: var(--green);
    }
    ~ .custom-control-label {
      color: var(--green);
      text-decoration: line-through;
    }
  }
  .custom-control-input:checked:disabled {
    ~ .custom-control-label::before {
      background-color: var(--gray);
      border-color: var(--gray);
    }
    ~ .custom-control-label {
      color: var(--gray);
    }
  }
}

.subtask-save-button {
  img {
    width: 15px;
  }
} 