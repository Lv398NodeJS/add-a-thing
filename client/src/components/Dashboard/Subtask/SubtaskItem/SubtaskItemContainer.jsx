import React, { Component } from 'react';
import SubtaskItem from './SubtaskItem';
import SubtaskEdit from './SubtaskEdit';

export class SubtaskItemContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditMode: false,
    };
  }

  toggleEditMode = () => {
    this.setState(prevState => ({
      isEditMode: !prevState.isEditMode,
    }));
  }

  render() {
    const {
      _id,
      name,
      completed,
      taskStatus,
    } = this.props;
    const { isEditMode } = this.state;
    return (
      isEditMode
        ? (
          <SubtaskEdit
            _id={_id}
            name={name}
            toggleEditMode={this.toggleEditMode}
          />
        )
        : (
          <SubtaskItem
            _id={_id}
            name={name}
            completed={completed}
            taskStatus={taskStatus}
            toggleEditMode={this.toggleEditMode}
          />
        )
    );
  }
}

export default SubtaskItemContainer;
