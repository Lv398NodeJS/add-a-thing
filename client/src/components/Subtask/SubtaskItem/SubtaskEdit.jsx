import React, { Component } from 'react';
import {
  Form, Row, Col, InputGroup, Button,
} from 'react-bootstrap';
import './SubtaskItem.scss';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import accept from '../../assets/accept.svg';
import * as subtaskActions from '../../../actions/subtaskActions';

class SubtaskEdit extends Component {
  handleSubmit = (event) => {
    const {
      _id,
      toggleEditMode,
      subtaskActions: { updateSubtask },
    } = this.props;
    event.preventDefault();
    event.stopPropagation();
    const name = this.input.value.trim().replace(/\s+/g, ' ');
    if (this.form.checkValidity() === false) {
      this.form.classList.add('was-validated');
    } else {
      this.form.classList.remove('was-validated');
      updateSubtask({ payload: name, key: 'name' }, _id);
    }
    toggleEditMode();
  }

  render() {
    const {
      name,
      _id,
    } = this.props;

    return (
      <Row className="subtask-row my-0 mx-0 px-2 rounded">
        <Col className="px-0">
          <Form
            _id={_id}
            className="needs-validation"
            method="POST"
            noValidate
            onSubmit={event => this.handleSubmit(event)}
            ref={(form) => {
              this.form = form;
            }}
          >
            <InputGroup>
              <Form.Control
                className="new-subtask-name was-validated"
                size="sm"
                minLength={1}
                maxLength={40}
                defaultValue={name}
                placeholder="Enter subtask name here..."
                required
                ref={(input) => {
                  this.input = input;
                }}
                onChange={(event) => {
                  if (!event.target.value.trim().length) event.target.value = '';
                }}
              />
              <InputGroup.Append>
                <Button
                  className="subtask-save-button"
                  variant="outline-primary"
                  size="sm"
                  type="submit"
                >
                  <img src={accept} alt="accept" />
                </Button>
              </InputGroup.Append>
              <div className="valid-feedback">Looks good!</div>
              <div className="invalid-feedback">Please provide subtask name.</div>
            </InputGroup>
          </Form>
        </Col>
      </Row>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  subtaskActions: bindActionCreators(subtaskActions, dispatch),
});
export { SubtaskEdit as SubtaskEditComponent };
export default connect(
  null,
  mapDispatchToProps,
)(SubtaskEdit);
