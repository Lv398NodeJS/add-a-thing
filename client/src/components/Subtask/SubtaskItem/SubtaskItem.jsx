import React, { Component } from 'react';
import {
  Form, Row, Col, Dropdown,
} from 'react-bootstrap';
import './SubtaskItem.scss';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as importedSubtaskActions from '../../../actions/subtaskActions';
import threedots from '../../assets/three-dots.svg';

class SubtaskItem extends Component {
  constructor(props) {
    super(props);
    const {
      taskRef,
      subtaskActions: { updateSubtask, deleteSubtask, convertToTask },
    } = this.props;
    this.taskRef = taskRef;
    this.updateSubtask = updateSubtask;
    this.deleteSubtask = deleteSubtask;
    this.convertToTask = convertToTask;
  }

  render() {
    const {
      name,
      completed,
      _id,
      taskStatus,
      toggleEditMode,
    } = this.props;

    return (
      <Row className="subtask-row my-0 mx-0 px-2 rounded">
        <Col
          className="d-flex justify-content-sm-start col-sm-11 px-0 my-1"
          onClick={() => taskStatus !== 'Done' && this.updateSubtask({ payload: !completed, key: 'completed' }, _id)}
        >
          <Form.Check type="checkbox" custom id={_id}>
            <Form.Check.Input
              type="checkbox"
              checked={completed}
              disabled={taskStatus === 'Done'}
              onChange={() => {
                this.updateSubtask({ payload: !completed, key: 'completed' }, _id);
              }}
            />
            <Form.Check.Label>{name}</Form.Check.Label>
          </Form.Check>
        </Col>
        <Col className="d-flex justify-content-sm-end col-sm-1 px-0">
          <Dropdown>
            <Dropdown.Toggle
              variant="light"
              size="sm"
              className="subtask-dropdown-toggle"
              disabled={taskStatus === 'Done'}
            >
              <img src={threedots} alt={threedots} className="threedots-icon" />
            </Dropdown.Toggle>

            <Dropdown.Menu>
              <Dropdown.Item onClick={() => toggleEditMode()}>
                Edit
              </Dropdown.Item>
              <Dropdown.Item onClick={() => this.deleteSubtask(_id)}>
                Delete
              </Dropdown.Item>
              <Dropdown.Item onClick={() => this.convertToTask(_id, { name }, document.URL.split('/').pop())}>
                Convert to Task
              </Dropdown.Item>
            </Dropdown.Menu>
          </Dropdown>
        </Col>
      </Row>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  subtaskActions: bindActionCreators(importedSubtaskActions, dispatch),
});
export { SubtaskItem as SubtaskItemComponent };
export default connect(
  null,
  mapDispatchToProps,
)(SubtaskItem);
