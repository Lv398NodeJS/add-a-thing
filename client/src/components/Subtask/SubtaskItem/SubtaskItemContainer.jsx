import React, { Component } from 'react';
import SubtaskItem from './SubtaskItem';

export class SubtaskItemContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditMode: false,
    };
  }

  toggleEditMode = () => {
    this.setState(prevState => ({
      isEditMode: !prevState.isEditMode,
    }));
  }

  render() {
    const {
      id,
      name,
      completed,
      taskStatus,
      taskRef,
    } = this.props;
    const { isEditMode } = this.state;
    return (
      isEditMode
        ? null
        : (
          <SubtaskItem
            id={id}
            name={name}
            completed={completed}
            taskStatus={taskStatus}
            taskRef={taskRef}
          />
        )
    );
  }
}

export default SubtaskItemContainer;
