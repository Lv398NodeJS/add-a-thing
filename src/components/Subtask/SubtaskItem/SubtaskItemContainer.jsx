import React, { Component } from 'react';
import SubtaskItem from './SubtaskItem';
import SubtaskEdit from './SubtaskEdit';

export class SubtaskListContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditMode: false,
    };
  }

  toggleEditMode = () => {
    this.setState(prevState => ({
      isEditMode: !prevState.isEditMode,
    }));
  }

  render() {
    const {
      id, text, completed, taskStatus, taskRef,
    } = this.props;
    const { isEditMode } = this.state;
    return (
      isEditMode
        ? (
          <SubtaskEdit
            taskRef={taskRef}
            id={id}
            text={text}
          />
        )
        : (
          <SubtaskItem
            id={id}
            text={text}
            completed={completed}
            taskStatus={taskStatus}
            taskRef={taskRef}
            toggleEditMode={this.toggleEditMode}
          />
        )
    );
  }
}

export default SubtaskListContainer;
