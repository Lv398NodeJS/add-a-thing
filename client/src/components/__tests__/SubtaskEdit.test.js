import React from 'react';
import { shallow } from 'enzyme';
import { SubtaskEditComponent } from '../Dashboard/Subtask/SubtaskItem/SubtaskEdit';

describe('SubtaskEdit component', () => {
  const subtaskActions = { editSubtaskText: jest.fn() };
  it('should render correctly and match the snapshot', () => {
    const subtaskEditComponent = shallow(
      <SubtaskEditComponent subtaskActions={subtaskActions} />,
    );
    expect(subtaskEditComponent).toMatchSnapshot();
  });


  it('should contain FormControl input with .new-subtask-name class name', () => {
    const subtaskEditComponent = shallow(
      <SubtaskEditComponent
        subtaskActions={subtaskActions}
      />,
    );
    expect(subtaskEditComponent.find('FormControl').hasClass('new-subtask-name')).toBeTruthy();
  });


  it('should contain FormControl input with placeholder "Enter subtask name here..."', () => {
    const subtaskEditComponent = shallow(
      <SubtaskEditComponent
        subtaskActions={subtaskActions}
      />,
    );
    expect(subtaskEditComponent.find('FormControl').prop('placeholder')).toBe('Enter subtask name here...');
  });

  it('should contain Button with .subtask-save-button class name', () => {
    const subtaskEditComponent = shallow(
      <SubtaskEditComponent
        subtaskActions={subtaskActions}
      />,
    );
    expect(subtaskEditComponent.find('Button').hasClass('subtask-save-button')).toBeTruthy();
  });

  it('should contain div with .valid-feedback class name', () => {
    const subtaskEditComponent = shallow(
      <SubtaskEditComponent
        subtaskActions={subtaskActions}
      />,
    );
    expect(subtaskEditComponent.find('.valid-feedback').length).toBe(1);
  });

  it('should contain div with .invalid-feedback class name', () => {
    const subtaskEditComponent = shallow(
      <SubtaskEditComponent
        subtaskActions={subtaskActions}
      />,
    );
    expect(subtaskEditComponent.find('.invalid-feedback').length).toBe(1);
  });
});
