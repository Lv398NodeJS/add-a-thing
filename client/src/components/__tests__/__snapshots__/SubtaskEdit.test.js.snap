// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubtaskEdit component should render correctly and match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SubtaskEdit
    subtaskActions={
      Object {
        "editSubtaskText": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <ForwardRef
        className="px-0"
      >
        <ForwardRef
          className="needs-validation"
          inline={false}
          method="POST"
          noValidate={true}
          onSubmit={[Function]}
        >
          <ForwardRef(render)>
            <ForwardRef
              autoFocus={true}
              className="new-subtask-name was-validated"
              maxLength={40}
              minLength={1}
              onChange={[Function]}
              placeholder="Enter subtask name here..."
              required={true}
              size="sm"
            />
            <ForwardRef>
              <ForwardRef
                active={false}
                className="subtask-save-button"
                disabled={false}
                size="sm"
                type="submit"
                variant="outline-primary"
              >
                <img
                  alt="Accept"
                  src="accept.svg"
                />
              </ForwardRef>
            </ForwardRef>
            <div
              className="valid-feedback"
            >
              Looks good!
            </div>
            <div
              className="invalid-feedback"
            >
              Please provide subtask name.
            </div>
          </ForwardRef(render)>
        </ForwardRef>
      </ForwardRef>,
      "className": "subtask-row my-0 mx-0 px-2 rounded",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <ForwardRef
          className="needs-validation"
          inline={false}
          method="POST"
          noValidate={true}
          onSubmit={[Function]}
        >
          <ForwardRef(render)>
            <ForwardRef
              autoFocus={true}
              className="new-subtask-name was-validated"
              maxLength={40}
              minLength={1}
              onChange={[Function]}
              placeholder="Enter subtask name here..."
              required={true}
              size="sm"
            />
            <ForwardRef>
              <ForwardRef
                active={false}
                className="subtask-save-button"
                disabled={false}
                size="sm"
                type="submit"
                variant="outline-primary"
              >
                <img
                  alt="Accept"
                  src="accept.svg"
                />
              </ForwardRef>
            </ForwardRef>
            <div
              className="valid-feedback"
            >
              Looks good!
            </div>
            <div
              className="invalid-feedback"
            >
              Please provide subtask name.
            </div>
          </ForwardRef(render)>
        </ForwardRef>,
        "className": "px-0",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "_id": undefined,
          "children": <ForwardRef(render)>
            <ForwardRef
              autoFocus={true}
              className="new-subtask-name was-validated"
              maxLength={40}
              minLength={1}
              onChange={[Function]}
              placeholder="Enter subtask name here..."
              required={true}
              size="sm"
            />
            <ForwardRef>
              <ForwardRef
                active={false}
                className="subtask-save-button"
                disabled={false}
                size="sm"
                type="submit"
                variant="outline-primary"
              >
                <img
                  alt="Accept"
                  src="accept.svg"
                />
              </ForwardRef>
            </ForwardRef>
            <div
              className="valid-feedback"
            >
              Looks good!
            </div>
            <div
              className="invalid-feedback"
            >
              Please provide subtask name.
            </div>
          </ForwardRef(render)>,
          "className": "needs-validation",
          "inline": false,
          "method": "POST",
          "noValidate": true,
          "onSubmit": [Function],
        },
        "ref": [Function],
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ForwardRef
                autoFocus={true}
                className="new-subtask-name was-validated"
                maxLength={40}
                minLength={1}
                onChange={[Function]}
                placeholder="Enter subtask name here..."
                required={true}
                size="sm"
              />,
              <ForwardRef>
                <ForwardRef
                  active={false}
                  className="subtask-save-button"
                  disabled={false}
                  size="sm"
                  type="submit"
                  variant="outline-primary"
                >
                  <img
                    alt="Accept"
                    src="accept.svg"
                  />
                </ForwardRef>
              </ForwardRef>,
              <div
                className="valid-feedback"
              >
                Looks good!
              </div>,
              <div
                className="invalid-feedback"
              >
                Please provide subtask name.
              </div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "autoFocus": true,
                "className": "new-subtask-name was-validated",
                "defaultValue": undefined,
                "maxLength": 40,
                "minLength": 1,
                "onChange": [Function],
                "placeholder": "Enter subtask name here...",
                "required": true,
                "size": "sm",
              },
              "ref": [Function],
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "Feedback": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "type": "valid",
                  },
                  "displayName": "Feedback",
                  "propTypes": Object {
                    "as": [Function],
                    "type": [Function],
                  },
                  "render": [Function],
                },
                "displayName": "FormControl",
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <ForwardRef
                  active={false}
                  className="subtask-save-button"
                  disabled={false}
                  size="sm"
                  type="submit"
                  variant="outline-primary"
                >
                  <img
                    alt="Accept"
                    src="accept.svg"
                  />
                </ForwardRef>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "active": false,
                  "children": <img
                    alt="Accept"
                    src="accept.svg"
                  />,
                  "className": "subtask-save-button",
                  "disabled": false,
                  "size": "sm",
                  "type": "submit",
                  "variant": "outline-primary",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "Accept",
                    "src": "accept.svg",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "active": false,
                    "disabled": false,
                    "type": "button",
                    "variant": "primary",
                  },
                  "displayName": "Button",
                  "render": [Function],
                },
              },
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "displayName": "InputGroupAppend",
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Looks good!",
                "className": "valid-feedback",
              },
              "ref": null,
              "rendered": "Looks good!",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Please provide subtask name.",
                "className": "invalid-feedback",
              },
              "ref": null,
              "rendered": "Please provide subtask name.",
              "type": "div",
            },
          ],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Append": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "InputGroupAppend",
              "render": [Function],
            },
            "Checkbox": [Function],
            "Prepend": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "InputGroupPrepend",
              "render": [Function],
            },
            "Radio": [Function],
            "Text": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "InputGroupText",
              "render": [Function],
            },
            "defaultProps": undefined,
            "displayName": "Bootstrap(InputGroup)",
            "propTypes": undefined,
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "Check": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Input": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "type": "checkbox",
              },
              "displayName": "FormCheckInput",
              "render": [Function],
            },
            "Label": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "type": "checkbox",
              },
              "displayName": "FormCheckLabel",
              "render": [Function],
            },
            "defaultProps": Object {
              "disabled": false,
              "inline": false,
              "isInvalid": false,
              "isValid": false,
              "title": "",
              "type": "checkbox",
            },
            "displayName": "FormCheck",
            "render": [Function],
          },
          "Control": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Feedback": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "type": "valid",
              },
              "displayName": "Feedback",
              "propTypes": Object {
                "as": [Function],
                "type": [Function],
              },
              "render": [Function],
            },
            "displayName": "FormControl",
            "render": [Function],
          },
          "Group": Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "FormGroup",
            "render": [Function],
          },
          "Label": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "column": false,
              "srOnly": false,
            },
            "displayName": "FormLabel",
            "render": [Function],
          },
          "Row": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "FormRow",
            "render": [Function],
          },
          "Text": Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "FormText",
            "render": [Function],
          },
          "defaultProps": Object {
            "inline": false,
          },
          "displayName": "Form",
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "Col",
        "render": [Function],
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": undefined,
      "displayName": "Bootstrap(Row)",
      "propTypes": undefined,
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <ForwardRef
          className="px-0"
        >
          <ForwardRef
            className="needs-validation"
            inline={false}
            method="POST"
            noValidate={true}
            onSubmit={[Function]}
          >
            <ForwardRef(render)>
              <ForwardRef
                autoFocus={true}
                className="new-subtask-name was-validated"
                maxLength={40}
                minLength={1}
                onChange={[Function]}
                placeholder="Enter subtask name here..."
                required={true}
                size="sm"
              />
              <ForwardRef>
                <ForwardRef
                  active={false}
                  className="subtask-save-button"
                  disabled={false}
                  size="sm"
                  type="submit"
                  variant="outline-primary"
                >
                  <img
                    alt="Accept"
                    src="accept.svg"
                  />
                </ForwardRef>
              </ForwardRef>
              <div
                className="valid-feedback"
              >
                Looks good!
              </div>
              <div
                className="invalid-feedback"
              >
                Please provide subtask name.
              </div>
            </ForwardRef(render)>
          </ForwardRef>
        </ForwardRef>,
        "className": "subtask-row my-0 mx-0 px-2 rounded",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ForwardRef
            className="needs-validation"
            inline={false}
            method="POST"
            noValidate={true}
            onSubmit={[Function]}
          >
            <ForwardRef(render)>
              <ForwardRef
                autoFocus={true}
                className="new-subtask-name was-validated"
                maxLength={40}
                minLength={1}
                onChange={[Function]}
                placeholder="Enter subtask name here..."
                required={true}
                size="sm"
              />
              <ForwardRef>
                <ForwardRef
                  active={false}
                  className="subtask-save-button"
                  disabled={false}
                  size="sm"
                  type="submit"
                  variant="outline-primary"
                >
                  <img
                    alt="Accept"
                    src="accept.svg"
                  />
                </ForwardRef>
              </ForwardRef>
              <div
                className="valid-feedback"
              >
                Looks good!
              </div>
              <div
                className="invalid-feedback"
              >
                Please provide subtask name.
              </div>
            </ForwardRef(render)>
          </ForwardRef>,
          "className": "px-0",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "_id": undefined,
            "children": <ForwardRef(render)>
              <ForwardRef
                autoFocus={true}
                className="new-subtask-name was-validated"
                maxLength={40}
                minLength={1}
                onChange={[Function]}
                placeholder="Enter subtask name here..."
                required={true}
                size="sm"
              />
              <ForwardRef>
                <ForwardRef
                  active={false}
                  className="subtask-save-button"
                  disabled={false}
                  size="sm"
                  type="submit"
                  variant="outline-primary"
                >
                  <img
                    alt="Accept"
                    src="accept.svg"
                  />
                </ForwardRef>
              </ForwardRef>
              <div
                className="valid-feedback"
              >
                Looks good!
              </div>
              <div
                className="invalid-feedback"
              >
                Please provide subtask name.
              </div>
            </ForwardRef(render)>,
            "className": "needs-validation",
            "inline": false,
            "method": "POST",
            "noValidate": true,
            "onSubmit": [Function],
          },
          "ref": [Function],
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <ForwardRef
                  autoFocus={true}
                  className="new-subtask-name was-validated"
                  maxLength={40}
                  minLength={1}
                  onChange={[Function]}
                  placeholder="Enter subtask name here..."
                  required={true}
                  size="sm"
                />,
                <ForwardRef>
                  <ForwardRef
                    active={false}
                    className="subtask-save-button"
                    disabled={false}
                    size="sm"
                    type="submit"
                    variant="outline-primary"
                  >
                    <img
                      alt="Accept"
                      src="accept.svg"
                    />
                  </ForwardRef>
                </ForwardRef>,
                <div
                  className="valid-feedback"
                >
                  Looks good!
                </div>,
                <div
                  className="invalid-feedback"
                >
                  Please provide subtask name.
                </div>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "autoFocus": true,
                  "className": "new-subtask-name was-validated",
                  "defaultValue": undefined,
                  "maxLength": 40,
                  "minLength": 1,
                  "onChange": [Function],
                  "placeholder": "Enter subtask name here...",
                  "required": true,
                  "size": "sm",
                },
                "ref": [Function],
                "rendered": null,
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "Feedback": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "defaultProps": Object {
                      "type": "valid",
                    },
                    "displayName": "Feedback",
                    "propTypes": Object {
                      "as": [Function],
                      "type": [Function],
                    },
                    "render": [Function],
                  },
                  "displayName": "FormControl",
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <ForwardRef
                    active={false}
                    className="subtask-save-button"
                    disabled={false}
                    size="sm"
                    type="submit"
                    variant="outline-primary"
                  >
                    <img
                      alt="Accept"
                      src="accept.svg"
                    />
                  </ForwardRef>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "active": false,
                    "children": <img
                      alt="Accept"
                      src="accept.svg"
                    />,
                    "className": "subtask-save-button",
                    "disabled": false,
                    "size": "sm",
                    "type": "submit",
                    "variant": "outline-primary",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "alt": "Accept",
                      "src": "accept.svg",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "img",
                  },
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "defaultProps": Object {
                      "active": false,
                      "disabled": false,
                      "type": "button",
                      "variant": "primary",
                    },
                    "displayName": "Button",
                    "render": [Function],
                  },
                },
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": undefined,
                  "displayName": "InputGroupAppend",
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Looks good!",
                  "className": "valid-feedback",
                },
                "ref": null,
                "rendered": "Looks good!",
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Please provide subtask name.",
                  "className": "invalid-feedback",
                },
                "ref": null,
                "rendered": "Please provide subtask name.",
                "type": "div",
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "Append": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "displayName": "InputGroupAppend",
                "render": [Function],
              },
              "Checkbox": [Function],
              "Prepend": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "displayName": "InputGroupPrepend",
                "render": [Function],
              },
              "Radio": [Function],
              "Text": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": undefined,
                "displayName": "InputGroupText",
                "render": [Function],
              },
              "defaultProps": undefined,
              "displayName": "Bootstrap(InputGroup)",
              "propTypes": undefined,
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "Check": Object {
              "$$typeof": Symbol(react.forward_ref),
              "Input": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "type": "checkbox",
                },
                "displayName": "FormCheckInput",
                "render": [Function],
              },
              "Label": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "type": "checkbox",
                },
                "displayName": "FormCheckLabel",
                "render": [Function],
              },
              "defaultProps": Object {
                "disabled": false,
                "inline": false,
                "isInvalid": false,
                "isValid": false,
                "title": "",
                "type": "checkbox",
              },
              "displayName": "FormCheck",
              "render": [Function],
            },
            "Control": Object {
              "$$typeof": Symbol(react.forward_ref),
              "Feedback": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "type": "valid",
                },
                "displayName": "Feedback",
                "propTypes": Object {
                  "as": [Function],
                  "type": [Function],
                },
                "render": [Function],
              },
              "displayName": "FormControl",
              "render": [Function],
            },
            "Group": Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "FormGroup",
              "render": [Function],
            },
            "Label": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "column": false,
                "srOnly": false,
              },
              "displayName": "FormLabel",
              "render": [Function],
            },
            "Row": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "FormRow",
              "render": [Function],
            },
            "Text": Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "FormText",
              "render": [Function],
            },
            "defaultProps": Object {
              "inline": false,
            },
            "displayName": "Form",
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "Col",
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": undefined,
        "displayName": "Bootstrap(Row)",
        "propTypes": undefined,
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
